sphericart-torch/src/torch_cuda_wrapper.cpp

```diff
@@ -0,0 +1,157 @@
+
+    auto find_num_blocks = [](int x, int bdim) {
+        return (x + bdim - 1) / bdim;
+    };
+
+    dim3 block_dim(find_num_blocks(xyz.size(0), GRID_DIM_Y));
+
+    switch (xyz.scalar_type()) {
+    case torch::ScalarType::Double:
+        sphericart::cuda::spherical_harmonics_cuda_base<double>(
+            xyz.data_ptr<double>(), xyz.size(0), prefactors.data_ptr<double>(),
+            prefactors.size(0), l_max, normalize, GRID_DIM_X, GRID_DIM_Y,
+            xyz.requires_grad(), gradients, hessian, sph.data_ptr<double>(),
+            d_sph.data_ptr<double>(), hess_sph.data_ptr<double>());
+        break;
+    case torch::ScalarType::Float:
+        sphericart::cuda::spherical_harmonics_cuda_base<float>(
+            xyz.data_ptr<float>(), xyz.size(0), prefactors.data_ptr<float>(),
+            prefactors.size(0), l_max, normalize, GRID_DIM_X, GRID_DIM_Y,
+            xyz.requires_grad(), gradients, hessian, sph.data_ptr<float>(),
+            d_sph.data_ptr<float>(), hess_sph.data_ptr<float>());
+        break;
+    }
+
+    cudaDeviceSynchronize();
+
+    if (!gradients)
+        d_sph = torch::Tensor();
+    if (!hessian)
+        hess_sph = torch::Tensor();
+    return {sph, d_sph, hess_sph};
+}
+
+/*
+    Torch wrapper for the CUDA kernel backwards pass.
+*/
+torch::Tensor sphericart_torch::spherical_harmonics_backward_cuda(
+    torch::Tensor xyz, torch::Tensor dsph, torch::Tensor sph_grad) {
+
+    if (!xyz.device().is_cuda()) {
+        throw std::runtime_error(
+            "internal error: CUDA version called on non-CUDA tensor");
+    }
+
+    auto xyz_grad = torch::Tensor();
+
+    if (xyz.requires_grad()) {
+        xyz_grad = torch::empty_like(xyz);
+
+        switch (xyz.scalar_type()) {
+        case torch::ScalarType::Double:
+            sphericart::cuda::spherical_harmonics_backward_cuda_base<double>(
+                dsph.data_ptr<double>(), sph_grad.data_ptr<double>(),
+                dsph.size(0), sph_grad.size(1), xyz_grad.data_ptr<double>());
+
+            break;
+        case torch::ScalarType::Float:
+            sphericart::cuda::spherical_harmonics_backward_cuda_base<float>(
+                dsph.data_ptr<float>(), sph_grad.data_ptr<float>(),
+                dsph.size(0), sph_grad.size(1), xyz_grad.data_ptr<float>());
+            break;
+        }
+
+        cudaDeviceSynchronize();
+    }
+
+    return xyz_grad;
+}
+
+/*
+    wrapper to compute prefactors with correct dtype.
+
+*/
+
+torch::Tensor sphericart_torch::prefactors_cuda(int64_t l_max,
+                                                at::ScalarType dtype) {
+    auto result =
+        torch::empty({(l_max + 1) * (l_max + 2)},
+                     torch::TensorOptions().device("cpu").dtype(dtype));
+
+    if (dtype == c10::kDouble) {
+        compute_sph_prefactors(l_max, static_cast<double *>(result.data_ptr()));
+    } else if (dtype == c10::kFloat) {
+        compute_sph_prefactors(l_max, static_cast<float *>(result.data_ptr()));
```

Comment:
I think `data_ptr` has a template parameter defaulting to `void`.


```suggestion
        compute_sph_prefactors(l_max, result.data_ptr<double>());
    } else if (dtype == c10::kFloat) {
        compute_sph_prefactors(l_max, result.data_ptr<float>());
```